### PR TITLE
Parallelize linting, and fix the build.

### PR DESCRIPTION
**Pull request by Irialad**
Linting large changes on a powerful machine is unnecessarily slow. This allows it to utilize all available cores.

### CODE REVIEW

File: gitlint/__init__.py [hunk 1]
```diff
@@ -163,6 +163,22 @@ def get_vcs_root():
     return (None, None)
 
 
+def processFile(filename, modified_files, commit, force, gitlint_config):
+    vcs, _ = get_vcs_root()
```
Comment (sk-) [inline gitlint/__init__.py]: This should be called only once, as it spawns two processes.

File: gitlint/__init__.py [hunk 2]
```diff
@@ -163,6 +163,22 @@ def get_vcs_root():
...
+
+    if force:
+        modified_lines = None
+    else:
+        modified_lines = vcs.modified_lines(filename,
+                                            modified_files[filename],
```
Comment (sk-) [inline gitlint/__init__.py]: Just pass the value corresponding to the filename

File: gitlint/__init__.py [hunk 3]
```diff
@@ -217,49 +233,52 @@ def main(argv, stdout=sys.stdout, stderr=sys.stderr):
...
-            stdout.write(output)
-            stdout.write(linesep + linesep)
+    args = [modified_files,
+            commit,
+            arguments['--force'],
+            gitlint_config]
+
+    with futures.ThreadPoolExecutor(max_workers=5) as executor:
+        future_to_file = {executor.submit(processFile, file, *args):
+                          file for file in sorted(modified_files.keys())}
```
Comment (sk-) [inline gitlint/__init__.py]: Please use filename instead.

File: gitlint/__init__.py [hunk 4]
```diff
@@ -217,49 +233,52 @@ def main(argv, stdout=sys.stdout, stderr=sys.stderr):
...
-        else:
-            output = linesep.join(output_lines)
-            stdout.write(output)
-            stdout.write(linesep + linesep)
+    args = [modified_files,
+            commit,
+            arguments['--force'],
+            gitlint_config]
+
+    with futures.ThreadPoolExecutor(max_workers=5) as executor:
```
Comment (sk-) [inline gitlint/__init__.py]: The number of workers should depend on the number of available cpus.

File: gitlint/__init__.py (hunk 3 above)
Comment (sk-) [inline gitlint/__init__.py]: Explicitly set the arguments here instead of using *args

**Comment by Irialad** [comment]
Before the change:
`$ time git lint -t
<SNIP>
real	5m0.441s
user	4m39.820s
sys	0m26.825s`

After the change:
`$ time git lint -t
<SNIP>
real	2m21.092s
user	7m52.605s
sys	0m33.318s`

File: gitlint/__init__.py (hunk 3 above)
Comment (Irialad) [reply gitlint/__init__.py]: Changed.

Comment (Irialad) [reply gitlint/__init__.py]: Ah, yes. Shadowing is bad form.

File: gitlint/__init__.py (hunk 4 above)
Comment (Irialad) [reply gitlint/__init__.py]: D'oh! Fixed.

File: gitlint/__init__.py (hunk 2 above)
Comment (Irialad) [reply gitlint/__init__.py]: Done.

File: gitlint/__init__.py (hunk 1 above)
Comment (Irialad) [reply gitlint/__init__.py]: Fixed.

**Comment by Irialad** [comment]
Travis-ci had a major outage. I've restarted the build of my fork. I don't have rights to restart this one.

**Comment by Irialad** [comment]
Pushed a noop change to retrigger the build.

File: setup.py [hunk 5]
```diff
@@ -37,6 +40,7 @@
         'scripts/custom_linters/tidy-wrapper.sh',
     ],
     install_requires=[
+        'futures',
```
Comment (sk-) [inline setup.py]: This should be installed only for python < 3
As explained in the futures package documentation, you should use the following line instead.
'futures; python_version == "2.7"'

File: gitlint/__init__.py [hunk 6]
```diff
@@ -44,6 +44,8 @@
 import os
 import os.path
 import sys
+from concurrent import futures
+from multiprocessing import cpu_count
```
Comment (sk-) [inline gitlint/__init__.py]: could you change this to just import the multiprocessing module. I follow the google style guide in that regards. https://google.github.io/styleguide/pyguide.html#Imports

File: gitlint/__init__.py [hunk 7]
```diff
@@ -217,49 +237,49 @@ def main(argv, stdout=sys.stdout, stderr=sys.stderr):
...
+                          filename for filename in sorted(modified_files.keys())
+                          }
+
+        for future in futures.as_completed(future_to_file):
+            result = future.result()
+            filename = future_to_file[future]
+            rel_filename = os.path.relpath(filename)
+
+            if not json_output:
+                stdout.write('Linting file: %s%s' %
```
Comment (sk-) [inline gitlint/__init__.py]: This is tricky. Now the message is no longer "Linting file ..." as the action already finished. Additionally you loose the ability to track whether the system is stuck in a particular file or not and that the files are returned in no consistent order (i.e. different runs will have different output).

An alternative is to use [executor.map](http://pythonhosted.org/futures/#concurrent.futures.Executor.map). That allows to keep the current semantics and actually simplifies the code as you can get rid of the `future_to_file` mapping.

File: gitlint/version.py [hunk 8]
```diff
@@ -0,0 +1 @@
+__VERSION__ = '0.0.9'
```
Comment (sk-) [inline gitlint/version.py]: Could you also add this change to the release notes in the readme file and if you want, give yourself credits for it.

**Comment by Irialad** [comment]
Thanks for the feedback. I will make changes, run tests, and update shortly.

File: gitlint/__init__.py (hunk 6 above)
Comment (Irialad) [reply gitlint/__init__.py]: Done.

File: gitlint/version.py (hunk 8 above)
Comment (Irialad) [reply gitlint/version.py]: Done.

File: setup.py (hunk 5 above)
Comment (Irialad) [reply setup.py]: Done.

**Comment by Irialad** [comment]
Another perf test:
Without the change at all:
real	2m4.373s
user	1m57.833s
sys	0m9.833s
With the new change, using map:
real	0m44.931s
user	2m39.182s
sys	0m10.156s

**Comment by Irialad** [comment]
The futures import using the 'newer' syntax apparently breaks python 3.3.

Would you prefer I fix this by using the older syntax? Or would you prefer to deprecate 3.3 support, and perhaps add 3.5 and 3.6?

**Comment by Irialad** [comment]
Is there anything at this point that I need to do to help drive this to release?

**Comment by sk-** [comment]
@Irialad I already uploaded a new version to pypi. Thanks again for the PR.

**Pull request merged**